src/parse-url-value.ts

```diff
@@ -0,0 +1,54 @@
+/**
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { TokenType, tokenize } from '@csstools/css-tokenizer'
+import {
+  type ComponentValue,
+  isFunctionNode,
+  isTokenNode,
+  parseListOfComponentValues,
+} from '@csstools/css-parser-algorithms'
+
+export const parseUrlValue = (string: string): string | null =>
+  parseListOfComponentValues(tokenize({ css: string }))
+    .map(
+      componentValue =>
+        parseUrlToken(componentValue) ?? parseUrlFunction(componentValue),
+    )
+    .find(Boolean) ?? null
+
+const parseUrlToken = (componentValue: ComponentValue): string | null => {
+  if (!isTokenNode(componentValue)) {
+    return null
+  }
+
+  const token = componentValue.value
+  return token[0] === TokenType.URL ? token[4].value : null
+}
+
+const parseUrlFunction = (componentValue: ComponentValue): string | null => {
+  if (!isFunctionNode(componentValue) || componentValue.getName() !== `url`) {
+    return null
+  }
+
+  for (const token of componentValue.tokens()) {
+    if (token[0] === TokenType.String) {
+      return token[4].value
+    }
+  }
+
+  return null
```

Comment:
@romainmenke hopefully last question for real haha, trying to reach 100% test coverage, but can't get this line covered. Is it actually impossible for this to be reached? Because if it's a `url(...)`, then it must either be a url function token or a url function with strings between it? Couldn't come up with input that gets to this part

If it's impossible, then I'll probably throw here and ignore this line for coverage

Reply:
Or I'll do something like this:
```ts
const parseUrlFunction = (componentValue: ComponentValue): string | null => {
  if (!isFunctionNode(componentValue) || componentValue.getName() !== `url`) {
    return null
  }

  return componentValue
    .tokens()
    .find((token): token is TokenString => token[0] === TokenType.String)![4]
    .value
}
```

Reply:
Yeah, it's impossible :)

https://www.w3.org/TR/2021/CRD-css-syntax-3-20211224/#consume-ident-like-token

There are three options :
- `url token`
- `bad url token` (we ignore those here, and that is fine)
- `function token` with a name of `url`

When the tokenizer algorithm encounters `"` or `'` after `url(` it will create a `function token` with a name of `url` and the next part will be either a `string token` or a `bad string token`.

A `bad string token` isn't possible because PostCSS will fail to parse those and throw syntax errors even before this plugin is invoked.

So you already have a guarantee that anything that isn't a `url token` wil be a `url function` with a valid string as it's first meaningful value.

Reply:
Makes sense, thanks for the explanation!